README.md
React Infinite Table
===

[![NPM version](https://img.shields.io/npm/v/react-infinite-table.svg?style=flat)](https://www.npmjs.com/package/react-infinite-table)
![NPM license](https://img.shields.io/npm/l/react-infinite-table.svg?style=flat)
[![npm bundle size](https://img.shields.io/bundlephobia/min/react-infinite-table)](https://bundlephobia.com/result?p=react-infinite-table@latest)
[![npm bundle size](https://img.shields.io/bundlephobia/minzip/react-infinite-table)](https://bundlephobia.com/result?p=react-infinite-table@latest)
[![NPM total downloads](https://img.shields.io/npm/dt/react-infinite-table.svg?style=flat)](https://npmcharts.com/compare/react-infinite-table?minimal=true)
[![NPM monthly downloads](https://img.shields.io/npm/dm/react-infinite-table.svg?style=flat)](https://npmcharts.com/compare/react-infinite-table?minimal=true)

**A browser-ready efficient scrolling table with fixed header and footer, and much more!** 

### Check out [the demo](http://nicoladefranceschi.github.io/react-infinite-table/)

## Features

- ✅ Render only the visibile rows  
- ✅ Fixed header
- ✅ Fixed footer
- ✅ Fixed left column(s)
- ✅ Column resize
- ✅ Column order changes (by dragging)
- ✅ Row selection (handling shift/modifier during click)
- ✅ Multiple rows selection
  * ⌘/Ctrl + Click: toggle selection
  * ⇧ + Click: select all the rows between the previously clicked row and the currently clicked row.
- 🔜 Rows with different heights / free height
- 🎉 Uses the HTML `<table>` standard tags!
- 🎉 No use of `javascript` to sync scroll between fixed columns/rows!


## Getting started

Install `react-infinite-table` using npm.

```shell
npm install react-infinite-table --save
```

Import and use as follows:

```jsx
import {Table} from 'react-infinite-table';

const rows = [
  { /* the row's data */ }, 
  //...
]

const columns = [
  { 
    cellRenderer: ({ columnIndex, column, rowData, rowIndex, className }) => {
      return (
        <td className={className}>
          {rowData.xxx}
        </td>
      )
    },
    headerRenderer: ({ columnIndex, column, className }) => {
      return (
        <th className={className}>
          {column.name}
        </th>
      )
    },
    footerRenderer: ({ columnIndex, column, className }) => {
      return (
        <td className={className}>
          {column.xxx}
        </td>
      )
    },
    width: 90,
    name: '...'
  },
  //...
]

<Table
  className='example-table'
  tableClassName='table table-bordered table-striped' // example using bootstrap
  height={200}
  rowHeight={30}
  rows={rows}
  columns={columns}
  headerCount={1}
  footerCount={1}
  fillTableWidth={true|false}
  rowIdKey='id'
  noRowsRenderer={() => 'No rows'}
  // keep the first column fixed:
  fixedColumnsCount={1}
  // row selection
  selectedRows={this.state.selectedRows}
  onSelectionChange={selectedRows => { this.setState({selectedRows}) }}
  canSelectMultipleRows={true|false}
  // infinite load:
  infiniteLoadBeginEdgeOffset={150}
  isInfiniteLoading={true|false}
  onInfiniteLoad={() => { /* fetch your data ... */ }}
  getLoadingSpinner={() => <div>Loading...</div>}
  // display from bottom upwards, like a Chat or Message Box 
  displayBottomUpwards={displayBottomUpwards}
  // allows column resize
  onColumnWidthChange={(columnIndex, width) => { /* update columns... */ }}
  // allows column reorder (by dragging)
  onColumnOrderChange={(fromIndex, toIndex) => { /* update columns... */ }}
/>
```


## Similar packages 

This package has been inspired by some existing library dealing with tables in `react`.

The initial code started from a "fork" of the [`react-infinite`](https://github.com/seatgeek/react-infinite) package, that implements a scrollable container with a lot of items by rendering only DOM nodes that the user is able to see (or might soon see).

The definition of the columns and the rendering of the cells is inspired by [`react-virtualized`](https://github.com/bvaughn/react-virtualized).

CSS is used to fix header, footer and some columns, using `position: sticky`. This has been inspired by [`ember-table`](https://github.com/Addepar/ember-table).

Comparison of similar packages:

| Feature        | **`react-infinite-table`** | [`react-virtualized`](https://github.com/bvaughn/react-virtualized) | [`react-base-table`](https://github.com/Autodesk/react-base-table) | 
| -------------- |:---:|:---:|:---:|
| HTML `<table>` standard tags | ✅ | ❌ | ❌ |
| Fixed columns  | ✅ | ❌ | ✅ |
| Fixed header   | ✅ | ✅ | ✅ |
| Fixed footer   | ✅ | ❌ | ❌ |
| Column resize (by dragging)  | ✅ | ❌ | ✅ |
| Column reorder (by dragging) | ✅ | ❌ | ❌ |
| Row selection  | ✅ | ❌ | ✅ |
| No use of `javascript` to sync scroll between fixed columns/rows | ✅ | ❌ | ❌ |
| Size | ![npm bundle size](https://img.shields.io/bundlephobia/min/react-infinite-table) <br/> ![npm bundle size](https://img.shields.io/bundlephobia/minzip/react-infinite-table) | ![npm bundle size](https://img.shields.io/bundlephobia/min/react-virtualized) <br/>  ![npm bundle size](https://img.shields.io/bundlephobia/minzip/react-virtualized) | ![npm bundle size](https://img.shields.io/bundlephobia/min/react-base-table) <br/> ![npm bundle size](https://img.shields.io/bundlephobia/minzip/react-base-table) |



## Dependencies

_react-infinite-table_ has very few dependencies that are managed by `npm`.

The following peer dependencies must be specified by your project in order to avoid version conflicts:
[`react`](https://www.npmjs.com/package/react),
[`react-dom`](https://www.npmjs.com/package/react-dom).
NPM will not automatically install these for you but it will show you a warning message with instructions on how to install them.


## Documentation

### Table Prop Types

`Table` is the main component and it's responsible of rendering your table.

| Property          | Type      | Required? | Description       |
| :---------------- | :-------- | :-------: | :---------------- |
| height            | Number    |     ✓     | The height of the table. If you want the table to take the height of the container, you should have a look at the [`AutoSizer` component](https://github.com/bvaughn/react-virtualized-auto-sizer) |
| overscanSize      | Number    |           | How much you want to render above/below the visible bounds of the table (in pixels). Default: `overscanSize: 500` |
| rowHeight         | Number    |     ✓     | A fixed row height _(TODO: planning to allow undefined, and each row will auto size to fit content!)_ |
| rows              | Array     |     ✓     | An array of data, one object for each row. There are no required keys, but if you want to use the "row selection" feature, an unique "row id" will be needed. |
| columns           | [Column](#column-prop-types) |     ✓     | One or more columns describing the data displayed in this table |
| headerCount       | Number    |           | Number of header rows in the `<thead>` section of the table. Default: `headerCount: 1` |
| footerCount       | Number    |           | Number of footer rows in the `<tfoot>` section of the table. Default: `footerCount: 0` |
| fixedColumnsCount | Number    |           | Number of columns to keep fixed on the left of the table while scrolling horizontally. Default: `fixedColumnsCount: 0` |
| fillTableWidth    | Boolean   |           | If `fillTableWidth = true` and the combined width of the columns is less than the table's width, the columns will grow in order to fill the table. Default: `fillTableWidth: false` |
| noRowsRenderer    | Function  |           | Callback used to render placeholder content when the number of rows is 0. Default: `noRowsRenderer: undefined` - the table will be empty. |
| rowIdKey          | String    |           | The key used to extract the id of a row from the row's data object. If present, it is used as a `key` for the table's row, otherwise the `rowIndex` is used. **Needed if row selection is enabled**. |
| selectedRows      | Object    |           | An object with the selected rows' ids as key, and `true` as value. Example: `{ 1: true, 2: true, 5: true }` - the rows with ids 1, 2 and 5 will be selected. Default: `selectedRows: undefined` - no selected rows. |
| canSelectMultipleRows | Boolean |         | Whether or not multiple rows can be selected. If true, using _⌘/Ctrl + Click_ on a row, the row will be selected/deselected, and using _⇧ + Click_ will select all the rows between the previously selected row and the clicked row. Default: `canSelectMultipleRows: false` |
| onSelectionChange | Function  |           | The callback to call when the selection changes. For example, use `(selectionRows) => { this.setState({ selectionRows }) }` to update your state to reflect the new selection. Default: `onSelectionChange: undefined` - Row selection is not allowed. |
| infiniteLoadBeginEdgeOffset | Number |    | How far from the bottom of the table we call the `onInfiniteLoad` function to fetch the new data. Default: `infiniteLoadBeginEdgeOffset: undefined` - Infinite load is not enabled. |
| onInfiniteLoad    | Function  |           | The callback used to start a fetch operation to get new rows. You should use this callback to start the async loading of new data and then update the `rows` props with the new rows. Default: `() => {}` |
| getLoadingSpinner | Function  |           | Callback used to render a loading message/spinner during data loading. Default: ` () => <div />` |
| isInfiniteLoading | Boolean   |           | Set to true if you want the show the loading spinner. Default: `isInfiniteLoading: undefined`  |
| displayBottomUpwards | Boolean |          | Whether or not to show the rows starting at the bottom, like in a chat/messaging application. Default: `displayBottomUpwards: false` |
| className         | String    |           | A class to add to the table container `<div>`. Default: `className: undefined` |
| tableClassName    | String    |           | A class to add to the `<table>` node. Default: `tableClassName: undefined` |
| style             | Object    |           | Passthrough prop to the table container `<div>`. Default: `style: undefined` |
| onColumnWidthChange | Function |          | Callback called when a column's width has changed. You should update the Column definition accordingly. `(columnIndex, width) => { /* update columns... */ }`. Default: `onColumnWidthChange: undefined` - Column's can't be resized. |
| onColumnOrderChange | Function |          | Callback called when a column order position is changed from `fromIndex` to `toIndex`. You should update the Columns array accordingly. `(fromIndex, toIndex) => { /* update columns... */ }`. Default: `onColumnOrderChange: undefined` - Column's can't be reordered. <br/> You can use the helper function `reorderColumns`. [Learn more](#reordercolumns) |


### Column Prop Types

| Property          | Type      | Required? | Description       |
| :---------------- | :-------- | :-------: | :---------------- |
| width             | Number    |     ✓     | The width of the column in pixels. |
| cellRenderer      | Function  |     ✓     | Callback responsible for rendering a cell's contents. [Learn more](#cellrenderer) |
| headerRenderer    | Function  |           | Optional callback responsible for rendering a column's header column. [Learn more](#headerrenderer) |
| footerRenderer    | Function  |           | Optional callback responsible for rendering a column's footer column. [Learn more](#footerrenderer) |


#### cellRenderer

Callback responsible for rendering a cell's contents.
It should implement the following signature:

```jsx
({
  className: string,
  columnIndex: number,
  column: object,
  rowIndex: number,
  rowData: object
}) => {
  return <td className={className}>{/*...*/}</td>
}
```

You should return a `<td>` node (or a `Component` that renders a `td`) since this node will be rendered in the `tbody > tr` section of the table.
You should always pass the `className` prop.


#### headerRenderer

Callback responsible for rendering a cell's header column.
It should implement the following signature:

```jsx
({
  className: string,
  columnIndex: number,
  column: object
}) => {
  return <th className={className}>{/*...*/}</th>
}
```

You should return a `<th>` node (or a `Component` that renders a `th`) since this node will be rendered in the `thead > tr` section of the table.
You should always pass the `className` prop.


#### footerRenderer

Callback responsible for rendering a cell's footer column.
It should implement the following signature:

```jsx
({
  className: string,
  columnIndex: number,
  column: object
}) => {
  return <td className={className}>{/*...*/}</td>
}
```

You should return a `<td>` node (or a `Component` that renders a `td`) since this node will be rendered in the `tfoot > tr` section of the table.
You should always pass the `className` prop.


### Utils

#### reorderColumns

This function is provided as an helper to update the `columns` array after the columns have been reorderer.

```jsx
import { reorderColumns } from 'react-infinite-table'

// ...

const newColumns = reorderColumns(oldColumns, fromIndex, toIndex)
```


### Style

You should import the `react-infinite-table/dist/style.css` file, or if you use `scss`, you can import the styles as follows:

```scss
@import "~react-infinite-table/src/style.scss";
```

If you want to use the "row selection" feature, you should apply a style on the `tr.tr-selected` rows as follows:

```css
.react-infinite-table tbody tr.tr-selected td {
  background-color: #007bff;
  border-color: #007bff;
  color: #ffffff;
}
```

If you use `bootstrap`, you need to apply the following fixes: 

```scss
@import "~bootstrap/scss/bootstrap.scss";

// For "Striped" Tables
// by default, bootstrap apply the background color to the <tr>
// and it's a semi-transparent color.

// we need an opaque color
$table-accent-bg: #F2F2F2;

// and we need to apply the backgrounds on td, not on tr!
.table-striped {
  tbody tr:nth-of-type(#{$table-striped-order}) {
    background-color: unset;
  }
  tbody tr:nth-of-type(#{$table-striped-order}) td {
    background-color: $table-accent-bg;
  }
}

// For "Bordered" Tables
// we need to apply the table border on the "wrapper" and not on 
// the <table> so that it is not hidden when scrolling.
.react-infinite-table {
  .table-bordered {
    border: none;
  }
  .react-infinite-table-wrapper {
    border: $table-border-width solid $table-border-color;
  }
}
```


## Development

You can run the demo: 

```shell
npm start
```

and make changes to the code. The browser will automatically load the changes.

## Contributions

Use [GitHub issues](https://github.com/nicoladefranceschi/react-infinite-table/issues) for requests.

I welcome pull requests. **Please ask first** before embarking on any significant pull request, otherwise you risk spending a lot of time working on something that the project's developers might not want to merge into the project.


## Changelog

Changes are tracked in the [changelog](https://github.com/nicoladefranceschi/react-infinite-table/blob/master/CHANGELOG.md).


## License

_react-infinite-table_ is available under the MIT License.


CHANGELOG.md
## Changelog

##### 1.3.0

- Remove `key` from renderer callbacks, the `key` is handled internally
- Changed the way React `key` is computed for `<tr>`s:
  - if the row was render before, use the same key
  - otherwise, try to reuse an existing row that was removed from the visible window

##### 1.2.3

- Added prop: use `fillTableWidth = true` to grow columns to fill the width of the table

##### 1.2.2

- Fix: handle scroll during column dragging

##### 1.2.1

- .npmignore

##### 1.2.0

- Remove dependency on `react-draggable`
- Refactoring

##### 1.1.2

- Bug fixes

##### 1.1.1

- Passthrough style prop to the table container `<div>`

##### 1.1.0

- Rename `fixedColumnsLeftCount` prop to `fixedColumnsCount`
- Documentation

##### 1.0.5

- Row selection

##### 1.0.4

- Fixes

##### 1.0.0

- Initial release.

src/demo/App.js
/* eslint-disable react/jsx-handler-names */
/* eslint-disable react/prop-types */

import React from 'react'

// Import library
import { Table, reorderColumns } from '../index.js'
import '../style.scss'

import './App.scss'

const N_ROWS = 10 ** 3
const INFINITE_SCROLLING_N_ROWS = 30
const N_COLS = 20

function cellRenderer ({
  columnIndex,
  column,
  rowData,
  rowIndex,
  className
}) {
  return (
    <td className={className}>
      R:{rowData.i} C:{column.i}
    </td>
  )
}

function headerRenderer ({
  columnIndex,
  column,
  className
}) {
  return (
    <th className={className}>
      C:{column.i}
    </th>
  )
}

function footerRenderer ({
  columnIndex,
  column,
  className
}) {
  return (
    <td className={className}>
      C:{column.i}
    </td>
  )
}

function createColumns (numberOfColumns) {
  const _columns = []

  for (let index = 0; index < numberOfColumns; index++) {
    _columns.push({
      i: index,
      cellRenderer: cellRenderer,
      headerRenderer: headerRenderer,
      footerRenderer: footerRenderer,
      width: 100
    })
  }
  return _columns
}

function createAllRows () {
  const rows = []
  for (let index = 0; index < N_ROWS; index++) {
    rows.push({ i: rows.length })
  }
  return rows
}

function recreateRows (infiniteScrolling, displayBottomUpwards) {
  let rows
  if (infiniteScrolling) {
    rows = []
    for (let index = 0; index < INFINITE_SCROLLING_N_ROWS; index++) {
      rows.push({ i: rows.length })
    }
  } else {
    rows = createAllRows()
  }
  if (displayBottomUpwards) {
    rows.reverse()
  }
  return rows
}

class App extends React.Component {
  state = {
    noRows: false,
    numberOfColumns: N_COLS,
    columns: createColumns(N_COLS),
    fixedColumnsCount: 2,
    fillTableWidth: false,
    displayBottomUpwards: false,
    infiniteScrolling: false,
    rows: createAllRows(),
    isInfiniteLoading: false,
    selectedRows: {},
    canSelectMultipleRows: true
  }

  componentWillUnmount () {
    clearTimeout(this._loadRowsTimeout)
  }

  onNoRowsChanged = noRows => {
    const rows = noRows ? [] : recreateRows(this.state.infiniteScrolling, this.state.displayBottomUpwards)
    this.setState({
      noRows: noRows,
      rows: rows
    })
  }

  onNumberOfColumnsChange = numberOfColumns => {
    this.setState({
      numberOfColumns: numberOfColumns,
      columns: createColumns(numberOfColumns)
    })
  }

  setDisplayBottomUpwards = displayBottomUpwards => {
    clearTimeout(this._loadRowsTimeout)
    const rows = recreateRows(this.state.infiniteScrolling, displayBottomUpwards)
    this.setState({
      noRows: false,
      displayBottomUpwards,
      rows: rows,
      isInfiniteLoading: false
    })
  }

  onFillTableWidth = fillTableWidth => {
    this.setState({
      fillTableWidth: fillTableWidth
    })
  }

  onInfiniteScrolling = infiniteScrolling => {
    clearTimeout(this._loadRowsTimeout)
    const rows = recreateRows(infiniteScrolling, this.state.displayBottomUpwards)
    this.setState({
      noRows: false,
      infiniteScrolling,
      rows: rows,
      isInfiniteLoading: false
    })
  }

  onInfiniteLoad = () => {
    console.log('Loading new rows!')
    this.setState({
      isInfiniteLoading: true
    })
    this._loadRowsTimeout = setTimeout(() => {
      const displayBottomUpwards = this.state.displayBottomUpwards
      const rows = [...this.state.rows]
      if (displayBottomUpwards) {
        rows.reverse()
      }

      for (let index = 0; index < INFINITE_SCROLLING_N_ROWS; index++) {
        rows.push({ i: rows.length })
      }

      if (displayBottomUpwards) {
        rows.reverse()
      }

      this.setState({
        rows: rows,
        isInfiniteLoading: false
      })
    }, 2000)
  }

  onFixedColumnsCountChange = fixedColumnsCount => {
    this.setState({
      fixedColumnsCount: fixedColumnsCount
    })
  }

  onColumnWidthChange = (columnIndex, width) => {
    const newColumns = [...this.state.columns]
    newColumns[columnIndex] = {
      ...newColumns[columnIndex],
      width: width
    }
    this.setState({
      columns: newColumns
    })
  }

  onColumnOrderChange = (fromIndex, toIndex) => {
    this.setState({
      columns: reorderColumns(this.state.columns, fromIndex, toIndex)
    })
  }

  onSelectionChange = selectedRows => {
    this.setState({
      selectedRows: selectedRows
    })
  }

  setCanSelectMultipleRows = canSelectMultipleRows => {
    this.setState({
      canSelectMultipleRows: canSelectMultipleRows
    })
  }

  render () {
    const {
      numberOfColumns,
      noRows,
      fixedColumnsCount,
      fillTableWidth,
      infiniteScrolling,
      displayBottomUpwards,
      rows,
      columns,
      selectedRows,
      canSelectMultipleRows
    } = this.state

    return (
      <div className='container'>
        <a href='https://github.com/nicoladefranceschi/react-infinite-table' className='float-right mt-3 mr-2'>
          <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 20 20' width='24' height='24'><title>GitHub</title><path d='M10 0a10 10 0 0 0-3.16 19.49c.5.1.68-.22.68-.48l-.01-1.7c-2.78.6-3.37-1.34-3.37-1.34-.46-1.16-1.11-1.47-1.11-1.47-.9-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.9 1.52 2.34 1.08 2.91.83.1-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.94 0-1.1.39-1.99 1.03-2.69a3.6 3.6 0 0 1 .1-2.64s.84-.27 2.75 1.02a9.58 9.58 0 0 1 5 0c1.91-1.3 2.75-1.02 2.75-1.02.55 1.37.2 2.4.1 2.64.64.7 1.03 1.6 1.03 2.69 0 3.84-2.34 4.68-4.57 4.93.36.31.68.92.68 1.85l-.01 2.75c0 .26.18.58.69.48A10 10 0 0 0 10 0' /></svg>
        </a>
        <h2>
          react-infinite-table
        </h2>
        <div className='App'>
          <div className='settings'>
            <div className='form-check'>
              <input
                className='form-check-input' type='checkbox'
                id='noRows'
                checked={noRows}
                onChange={e => this.onNoRowsChanged(e.target.checked)}
              />
              <label className='form-check-label' htmlFor='noRows'>
                No rows
              </label>
            </div>
            <div className='form-inline'>
              <label className='mr-2' htmlFor='numberOfColumns'>Number of columns: </label>
              <input
                className='form-control'
                type='number'
                id='numberOfColumns'
                value={numberOfColumns}
                min={1}
                step={1}
                onChange={e => this.onNumberOfColumnsChange(parseInt(e.target.value) || N_COLS)}
              />
            </div>
            <div className='form-inline'>
              <label className='mr-2' htmlFor='fixedColumnsCount'>Fixed columns: </label>
              <input
                className='form-control'
                type='number'
                id='fixedColumnsCount'
                value={fixedColumnsCount}
                min={0}
                max={columns.length}
                step={1}
                onChange={e => this.onFixedColumnsCountChange(parseInt(e.target.value) || 0)}
              />
            </div>
            <div className='form-check'>
              <input
                className='form-check-input' type='checkbox'
                id='fillTableWidth'
                checked={fillTableWidth}
                onChange={e => this.onFillTableWidth(e.target.checked)}
              />
              <label htmlFor='fillTableWidth'> Fill table width</label>
            </div>
            <div className='form-check'>
              <input
                className='form-check-input' type='checkbox'
                id='infiniteScrolling'
                checked={infiniteScrolling}
                onChange={e => this.onInfiniteScrolling(e.target.checked)}
              />
              <label htmlFor='infiniteScrolling'> Infinite scrolling</label>
            </div>
            <div className='form-check'>
              <input
                className='form-check-input' type='checkbox'
                id='displayBottomUpwards'
                checked={displayBottomUpwards}
                onChange={e => this.setDisplayBottomUpwards(e.target.checked)}
              />
              <label htmlFor='displayBottomUpwards'> Display Bottom Upwards</label>
            </div>
            <div className='form-check'>
              <input
                className='form-check-input' type='checkbox'
                id='canSelectMultipleRows'
                checked={canSelectMultipleRows}
                onChange={e => this.setCanSelectMultipleRows(e.target.checked)}
              />
              <label htmlFor='canSelectMultipleRows'> Can select multiple rows</label>
            </div>
          </div>
          <Table
            className='example-table'
            tableClassName='table table-bordered table-striped'
            height={600}
            rowHeight={50}
            rows={rows}
            columns={columns}
            fixedColumnsCount={fixedColumnsCount}
            headerCount={1}
            footerCount={1}
            fixBorders
            fillTableWidth={fillTableWidth}
            noRowsRenderer={() => 'No rows'}
            rowIdKey='i'
            selectedRows={selectedRows}
            canSelectMultipleRows={canSelectMultipleRows}
            onSelectionChange={this.onSelectionChange}
            infiniteLoadBeginEdgeOffset={infiniteScrolling ? 150 : undefined}
            isInfiniteLoading={infiniteScrolling ? this.state.isInfiniteLoading : undefined}
            onInfiniteLoad={infiniteScrolling ? this.onInfiniteLoad : undefined}
            getLoadingSpinner={() => <div>Loading...</div>}
            displayBottomUpwards={displayBottomUpwards}
            onColumnWidthChange={this.onColumnWidthChange}
            onColumnOrderChange={this.onColumnOrderChange}
          />
        </div>
      </div>
    )
  }
}

export default App
